core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/io/ContentLengthAwareSubscriber.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.internal.signer.io;
+
+import java.nio.ByteBuffer;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * Decorator subscriber that limits the number of bytes sent to the wrapped subscriber to at most {@code contentLength}. Once
+ * the given content length is reached, the upstream subscription is cancelled, and the wrapped subscriber is completed.
+ */
+@SdkInternalApi
+public class ContentLengthAwareSubscriber implements Subscriber<ByteBuffer> {
+    private final Subscriber<? super ByteBuffer> subscriber;
+    private Subscription subscription;
+    private boolean subscriptionCancelled;
+    private long remaining;
+
+    public ContentLengthAwareSubscriber(Subscriber<? super ByteBuffer> subscriber, long contentLength) {
+        this.subscriber = subscriber;
+        this.remaining = contentLength;
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        if (subscription == null) {
+            throw new NullPointerException("subscription must not be null");
+        }
+        this.subscription = subscription;
+        subscriber.onSubscribe(subscription);
+    }
+
+    @Override
+    public void onNext(ByteBuffer byteBuffer) {
+        if (remaining > 0) {
+            long bytesToRead = Math.min(remaining, byteBuffer.remaining());
+            // cast is safe, min of long and int is <= max_int
+            byteBuffer.limit(byteBuffer.position() + (int) bytesToRead);
+            remaining -= bytesToRead;
+            subscriber.onNext(byteBuffer);
+        } else if (remaining == 0 && !subscriptionCancelled) {
```

Comment:
Should we remove `else` so that we call `onComplete()` right away if there's no remaining?

Reply:
Makes sense!